### DISCUSSION TITLE
With C# 11 'ref fields', support 'ref local' and 'ref', 'out', or in parameter inside an anonymous method, lambda expression, query expression, or local function

### DISCUSSION THREAD

**Discussion by kasajian** [dotnet/csharplang#6895] [Language Ideas]
Currently there are two errors that are generated when you attempt to use a 'ref local' and 'ref', 'out' or 'in' parameters inside a lambda:

error CS8175: Cannot use ref local 'rx' inside an anonymous method, lambda expression, or query expression
error CS1628: Cannot use ref, out, or in parameter 'v' inside an anonymous method, lambda expression, query expression, or local function

The reason for this is because of the internal classes that are created by the C# compiler that copies that method parameters to class fields.   And fields cannot be 'ref' prior to C# 11.

But with C# 11, ref fields can be used.   This would be very useful for structs, such as ref locals being using of lambdas, otherwise unnecessary copying is performend.

Example code that demonstrates the compiler errors:
```
using System;
using System.Linq;

// Demonstrates CS8175 and CS1628

struct V
{
    public int i1;
}

// See comments for enabling errors.
public class C_demos_errors {
    public void M() {

        var local_v = new V();
        ref var refLocal_v = ref local_v;
        
        local_v.i1 = 34;
        
        var x1 = new object[1].Select(_ => _ is not null && local_v.i1 == 0);

        // Generates error CS8175: Cannot use ref local 'refLocal_v' inside an anonymous method, lambda expression, or query expression
        // var x2 = new object[1].Select(_ => _ is not null && refLocal_v.i1 == 0);
        
        var x3 = foo_byval(local_v);

        var x4 = foo_byref(ref local_v);
    }
    
    private Object foo_byval(V v) {
        return new object[1].Select(_ => _ is not null && v.i1 == 0);
    }

    private Object foo_byref(ref V v) {
        // Generates error CS1628: Cannot use ref, out, or in parameter 'v' inside an anonymous method, lambda expression, query expression, or local function
        // return new object[1].Select(_ => _ is not null && v.i1 == 0);
        return null;
    }
}
```
online example: [link](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwHQBkCWAdgI4DcyyA9JegCICmAtgPaEDOALgE4CGH9bdAGEAygA44AdgCs6HoWDCRcAGwwxFJJy4BXAMYd0ANWQBvZOkuYAzOiKH8cckgC+m6uhH166Pc0aM9IQcggBmzFzoQTwARgA2RADmUVxcEWzYyFC2sMIA+sBMzGx59Knp6OZIVjaYACzoALIAFACUlZo1NQBuPJFxzHo8cXnd6AC86IT0AO7Gbc5dVlz0oei9kSuhuIPDoxPoW+gDQyPdi0voFpcne93YjgfWdRdL10sb6AAecAfTc8wYgAregGADacAAutgvHFQRxmnkJgA+dBI/CCQjMQyEHRxOLoABkhOOuzOD1+40mqFazneXQ8AHEgmU+AIUmlIqIJDIQMJ5FjDDo2D4jrcCQByLY7U6jCV2dj4QpyQgq1gATxYwvQgQ4AAtmMAADTHHiMGLAHhRL4ABxWbDY+FYJoi6GIOjK6utdoEjtY9JqHk+Xxgf1m6EBIPBUJh9DhBkRKLRdkx2KmeIJxMOqxldwpE2ptIDVmLlmDtkm4WYeRi6t6cWa4tGRaQpfWfW+DUrzGrta2zTFZObr3QbmqJfHljt+F6/HQAHlgfD0FWa3Xhs0jOt2lVLphJFNw5H4RDobD4Ynxqj0amcRmiST7o8qegaSOx/Tp7OfIuo4ZV32qwDqsxjbh0k4MjQzLTLw/CCGUnJKKo6h8kIAppsKoqrC6OgcC6kREOgNp9Ga9D8JEErdPKRCOsq8hqoQmrMNquoGsaprmpa3r2n6hAmu6nrcb6Tp8RG/RkiuOiEAYIlth4UAHv8EZLtGZ5xheSJXsmGJTGmuL4g+6z5i+b5tgp6b4u+yAfq2tlAA===)

**Comment by Enderlook** [comment]
I don't think it would be possible to use `ref locals` in lambdas because they heap allocated delegates, and `ref fields` can only be used on `ref struct` (which are stack allocated).

Being allowed to use `ref locals` (or `ref structs`, which for this and the previous purpose are similar) in local methods would be nice but it produces the similar problem. If someone creates a delegate that points towards that local function it would be heap allocating it, which would be illegal. The compiler would require to check if a function has or not `ref locals` and produce a proper error when have one, which for newer C# users may be difficult to understand why some local methods can be turned into delegates but others can't.
Regardless of the possibility of confussion in users, I must say that personally I would love this feature. May times I found wanting to use a `ref local` or wanting to use a local `Span<T>` in a local method and having to add a parameter for it.
